### PR TITLE
[Client] Fix __getattr__ behavior for ClientActorHandles

### LINKED ISSUE #31581: [Client]  Direct serialization of an actor handle created under Ray Client fails
### What happened + What you expected to happen

https://discuss.ray.io/t/ray-cloudpickle-error-while-pickling-ray-dataset/5396/9

### Versions / Dependencies

master

### Reproduction script

```
import ray
import ray.cloudpickle as cloudpickle
ray.init("ray://localhost:10001")

@ray.remote
class Foo:
    def ping(self):
        return "ok"

f = Foo.remote()
s = cloudpickle.dumps(f)  # This raises the same error.
```

### Issue Severity

None

### PR DESCRIPTION
**Pull request by ckw017**
## Why are these changes needed?

Closes #31581

This is currently breaking serialization for client actor handles, which in turn breaks serialization for ray datasets. It also causes confusing error messages when accessing nonexistent fields on actor handles.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tests/test_client.py [hunk 1]
```diff
@@ -852,6 +853,54 @@ def test_ignore_reinit(call_ray_start_shared, shutdown_only):
     assert ctx1 == ctx2
 
 
+def test_client_actor_missing_field(call_ray_start_shared):
+    """
+    Test remote calls with large (multiple chunk) arguments
```
Comment (ijrsvt) [inline python/ray/tests/test_client.py:858]: Can you update this description?

File: python/ray/util/client/common.py [hunk 2]
```diff
@@ -473,14 +473,23 @@ def _actor_id(self) -> ClientActorRef:
         return self.actor_ref
 
     def __getattr__(self, key):
+        if key == "_method_num_returns":
+            # We need to explicitly handle this value since it is used below,
+            # otherwise we may end up infinitely recursing when deserializing.
+            # This can happen after unpickling an object but before
+            # _method_num_returns is correctly populated.
```
Comment (ijrsvt) [inline python/ray/util/client/common.py:480]: ooc, is this tested at all?

File: python/ray/util/client/common.py [hunk 3]
```diff
@@ -473,14 +473,23 @@ def _actor_id(self) -> ClientActorRef:
...
-            self._method_signatures.get(key),
-        )
+        if key in self._method_signatures:
+            return ClientRemoteMethod(
+                self,
+                key,
+                self._method_num_returns.get(key),
+                self._method_signatures.get(key),
+            )
+        raise AttributeError(f"ClientActorRef has no attribute '{key}'")
```
Comment (ijrsvt) [inline python/ray/util/client/common.py:492]:
```suggestion
        if key not in self._method_signatures:
            raise AttributeError(f"ClientActorRef has no attribute '{key}'")
        return ClientRemoteMethod(
            self,
            key,
            self._method_num_returns.get(key),
            self._method_signatures.get(key),
        )
```

**Review by ijrsvt — APPROVED** [review]
Overall looks good, a few minor comments

File: python/ray/util/client/common.py (hunk 2 above)
Comment (ckw017) [reply python/ray/util/client/common.py:480]: It's indirectly tested by the deserialization in the unit test, without this part it runs into the recursion limit. But beyond that, no

Comment (ijrsvt) [reply python/ray/util/client/common.py:480]: Sounds good--I just wanted to check that it was implicitly tested.

**Pull request merged**